src/legacy/ui/public/vis/vis_types/vislib_vis_type.js

```diff
@@ -83,12 +90,11 @@ export function VislibVisTypeProvider(Private, $rootScope, $timeout, $compile) {
           this.$scope.$digest();
         }
 
-        this.vis.vislibVis = new vislib.Vis(this.chartEl, this.vis.params);
-        this.vis.vislibVis.on('brush', this.vis.API.events.brush);
-        this.vis.vislibVis.on('click', this.vis.API.events.filter);
-        this.vis.vislibVis.on('renderComplete', resolve);
         this.vis.vislibVis.render(esResponse, this.vis.getUiState());
 
+        // refreshing the legend after the chart is rendered.
+        // this is necessary because some visualizations (heatmap and gauge)
+        // provide data necessary for the legend only after a render cycle.
         if (this.vis.params.addLegend) {
```

Comment:
do this only for heatmap and gauge.
also, it seems that for heatmap and gauge we also need to rerender the chart again to fix top/bottom legend overflow

Reply:
> do this only for heatmap and gauge

Does it make sense to make this code specific to certain chart types? If a plugin vis uses the same "hack", its legend wouldn't show up at all.

 > also, it seems that for heatmap and gauge we also need to rerender the chart again to fix top/bottom legend overflow

Yes, see my comment above regarding that. It's really edge-casey to trigger this bug for heatmap and gauge and I wasn't sure whether it makes sense to hurt performance for all heatmaps and gauges for this. If you think that's reasonable I will add a re render just for these chart types.

Reply:
if a third party plugin would want to do something similar, they would need to modify the vislib legend anyway as heatmap and gauge special legend handling is hardcoded in there.

 can we try offload the decision if to rerender or not to the legend itself ?

i agree hitting the bug with heatmap or gauge is less likely, but on dashboard it might still happen quite often. And since the solution is anyway a hack, it would be nice if it actually fully fixes a problem.

Reply:
> if a third party plugin would want to do something similar, they would need to modify the vislib legend anyway as heatmap and gauge special legend handling is hardcoded in there.

Ah, you are right. I moved the list of "special" visualizations to a central constant to emphasise the connection.